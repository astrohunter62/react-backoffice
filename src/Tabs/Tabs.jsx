import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from 'material-ui/styles'
import AppBar from 'material-ui/AppBar'
import MaterialTabs, { Tab } from 'material-ui/Tabs'
import Typography from 'material-ui/Typography'

function TabContainer(props) {
  return (
    <Typography component="div" style={{ padding: 8 * 3 }}>
      {props.children}
    </Typography>
  )
}

TabContainer.propTypes = {
  children: PropTypes.node.isRequired,
}

const styles = theme => ({
  root: {
    flexGrow: 1,
    width: '100%',
    marginTop: theme.spacing.unit * 3,
    backgroundColor: theme.palette.background.paper,
  },
})

class Tabs extends React.Component {
  static propTypes = {
    scrollable: PropTypes.bool,
    data: PropTypes.arrayOf(PropTypes.object).isRequired,
    classes: PropTypes.objectOf(PropTypes.string).isRequired,
  }

  static defaultProps = {
    scrollable: false,
  }

  state = {
    value: 0,
  };

  handleChange = (event, value) => {
    this.setState({ value })
  };

  render() {
    const { scrollable, data, classes } = this.props
    const { value } = this.state

    return (
      <div className={classes.root}>
        <AppBar position="static">
          <MaterialTabs
            value={value}
            onChange={this.handleChange}
            indicatorColor="secondary"
            scrollable={scrollable}
            scrollButtons="auto"
          >
            {data.map(item => (
              <Tab
                key={`tab-${(Math.random() * 100).toFixed(2)}`}
                label={item.title}
              />
            ))}
          </MaterialTabs>
        </AppBar>

        {data.map((item, index) => {
          if (value === index) {
            return (
              <TabContainer
                key={`tabcontainer-${(Math.random() * 100).toFixed(2)}`}
              >
                {item.content}
              </TabContainer>
            )
          }

          return null
        })}
      </div>
    )
  }
}

export default withStyles(styles)(Tabs)
