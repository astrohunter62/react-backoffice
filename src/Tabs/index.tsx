import React, { FunctionComponent } from "react";
import { withRouter } from "react-router-dom";
import {
  AppBar,
  withStyles,
  Tabs as MaterialTabs,
  Tab,
  Typography,
} from "@material-ui/core";

const TabContainer: FunctionComponent<any> = (props) => {
  return (
    <Typography component="div" style={{ padding: 8 * 3 }}>
      {props.children}
    </Typography>
  );
};

const styles = (theme: any) => ({
  root: {
    flexGrow: 1,
    width: "100%",
    marginTop: theme.spacing() * 3,
    backgroundColor: theme.palette.background.paper,
  },
});

type TabsProps = {
  isScrollable?: boolean;
  data: Record<string, any>[];
  classes: {
    [key: string]: string;
  };
};

type TabsState = {
  value?: number;
};

class Tabs extends React.Component<TabsProps, TabsState> {
  constructor(props: TabsProps) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
  }

  state = {
    value: 0,
  };

  componentDidMount() {
    const { data } = this.props;
    const hash = window.location.hash.replace("#/", "");
    let value = 0;

    if (data && data.constructor === Array) {
      data.forEach((item, index) => {
        if (item.id === hash) {
          value = index;
        }
      });
    }

    this.setState({
      value,
    });
  }

  handleChange(event: any, value: any) {
    const hash = this.props.data[value].id || value;
    window.location.hash = `#/${hash}`;
    this.setState({ value });
  }

  render() {
    const { isScrollable, data, classes } = this.props;
    const { value } = this.state;
    return (
      <div className={classes.root}>
        <AppBar position="static">
          <MaterialTabs
            value={value}
            onChange={this.handleChange}
            indicatorColor="secondary"
            variant={isScrollable ? "scrollable" : undefined}
            scrollButtons="auto"
          >
            {data.map((item, index) => (
              <Tab key={`tab-${index}`} label={item.title} />
            ))}
          </MaterialTabs>
        </AppBar>

        {data.map((item, index) => {
          if (value === index) {
            return (
              <TabContainer key={`tabcontainer-${index}`}>
                {item.content}
              </TabContainer>
            );
          }
          return null;
        })}
      </div>
    );
  }
}
const tabsWithStyles = withStyles(styles)(Tabs);

export default withRouter(tabsWithStyles as any) as any;
