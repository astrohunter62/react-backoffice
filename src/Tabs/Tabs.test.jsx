import React from 'react'
import MockRouter from 'react-mock-router'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { Tab } from '@material-ui/core'

import Tabs from '.'
import tabsContent from '../tests/data/tabs'

Enzyme.configure({ adapter: new Adapter() })

describe('Tabs', () => {
  it('renders correctly', () => {
    const tree = shallow((
      <MockRouter>
        <Tabs
          data={[{
            title: 'Title',
            content: (
              <p>Content</p>
            ),
          }]}
        />
      </MockRouter>
    ))

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly if no data', () => {
    const tree = shallow((
      <MockRouter>
        <Tabs data={[]} />
      </MockRouter>
    ))

    expect(tree).toMatchSnapshot()
  })

  it('click second tab', () => {
    const tabAt = 1
    const tabs = shallow(<MockRouter><Tabs data={tabsContent} /></MockRouter>)

    expect(window.location.hash).toBe('')
    tabs.find(Tab).at(tabAt).simulate('click')
    expect(window.location.hash).toBe(`#/${tabsContent[tabAt].id}`)
  })

  it('test if correct element is active', () => {
    const tabAt = 1
    window.location.hash = `#/${tabsContent[tabAt].id}`

    shallow(<MockRouter><Tabs data={tabsContent} /></MockRouter>)

    expect(window.location.hash).toBe(`#/${tabsContent[tabAt].id}`)
  })
})
