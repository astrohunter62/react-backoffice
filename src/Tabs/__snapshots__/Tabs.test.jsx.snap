// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders correctly 1`] = `
<div
  className="Tabs-root-1"
>
  <header
    className="MuiPaper-root-10 MuiPaper-elevation4-16 MuiAppBar-root-2 MuiAppBar-positionStatic-6 MuiAppBar-colorPrimary-8"
  >
    <div
      className="MuiTabs-root-37"
    >
      <div
        className="MuiTabs-flexContainer-38"
      >
        <div
          className="MuiTabs-scroller-39 MuiTabs-fixed-40"
          onScroll={[Function]}
          role="tablist"
          style={
            Object {
              "marginBottom": 0,
            }
          }
        >
          <div
            className="MuiTabs-flexContainer-38"
          >
            <button
              aria-selected={true}
              className="MuiButtonBase-root-62 MuiTab-root-46 MuiTab-textColorInherit-48 MuiTab-textColorInheritSelected-55"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="tab"
              tabIndex="0"
              type="button"
            >
              <span
                className="MuiTab-wrapper-58"
              >
                <span
                  className="MuiTab-labelContainer-59"
                >
                  <span
                    className="MuiTab-label-60"
                  >
                    Title
                  </span>
                </span>
              </span>
              <span
                className="MuiTouchRipple-root-68"
              />
            </button>
          </div>
          <span
            className="MuiTabIndicator-root-65 MuiTabIndicator-colorSecondary-67 MuiTabs-indicator-45"
            style={
              Object {
                "left": 0,
                "width": 0,
              }
            }
          />
        </div>
      </div>
    </div>
  </header>
  <div
    className="MuiTypography-root-75 MuiTypography-body1-84"
    style={
      Object {
        "padding": 24,
      }
    }
  >
    <p>
      Content
    </p>
  </div>
</div>
`;

exports[`Tabs renders correctly if no data 1`] = `
<div
  className="Tabs-root-1"
>
  <header
    className="MuiPaper-root-10 MuiPaper-elevation4-16 MuiAppBar-root-2 MuiAppBar-positionStatic-6 MuiAppBar-colorPrimary-8"
  >
    <div
      className="MuiTabs-root-37"
    >
      <div
        className="MuiTabs-flexContainer-38"
      >
        <div
          className="MuiTabs-scroller-39 MuiTabs-fixed-40"
          onScroll={[Function]}
          role="tablist"
          style={
            Object {
              "marginBottom": 0,
            }
          }
        >
          <div
            className="MuiTabs-flexContainer-38"
          />
          <span
            className="MuiTabIndicator-root-65 MuiTabIndicator-colorSecondary-67 MuiTabs-indicator-45"
            style={
              Object {
                "left": 0,
                "width": 0,
              }
            }
          />
        </div>
      </div>
    </div>
  </header>
</div>
`;
